lighthouse-viewer/app/src/main.js

```diff
@@ -0,0 +1,515 @@
+  }
+
+  /**
+   * Shows a butter bar.
+   * @param {!string} msg The message to show.
+   * @param {boolean=} autoHide True to hide the message after a duration.
+   *     Default is true.
+   */
+  log(msg, autoHide = true) {
+    clearTimeout(this._id);
+
+    this.el.textContent = msg;
+    this.el.classList.add('show');
+    if (autoHide) {
+      this._id = setTimeout(_ => {
+        this.el.classList.remove('show');
+      }, 7000);
+    }
+  }
+
+  warn(msg) {
+    console.warn(msg);
+  }
+
+  error(msg) {
+    console.error(msg);
+  }
+
+  /**
+   * Explicitly hides the butter bar.
+   */
+  hide() {
+    clearTimeout(this._id);
+    this.el.classList.remove('show');
+  }
+}
+
+const logger = new Logger('#log');
+
+/**
+ * Wrapper for Firebase Authentication
+ * @class
+ */
+class FirebaseAuth {
+  constructor() {
+    this.accessToken = null;
+    this.user = null;
+
+    this.provider = new firebase.auth.GithubAuthProvider();
+    this.provider.addScope('gist');
+
+    firebase.initializeApp({
+      apiKey: 'AIzaSyApMz8FHTyJNqqUtA51tik5Mro8j-2qMcM',
+      authDomain: 'lighthouse-viewer.firebaseapp.com',
+      databaseURL: 'https://lighthouse-viewer.firebaseio.com',
+      storageBucket: 'lighthouse-viewer.appspot.com',
+      messagingSenderId: '962507201498'
+    });
+
+    // Wrap auth state callback in a promise so  other parts of the app that
+    // require login can hook into the changes.
+    this.ready = new Promise((resolve, reject) => {
+      firebase.auth().onAuthStateChanged(user => {
+        idb.get('accessToken').then(token => {
+          if (user && token) {
+            this.accessToken = token;
+            this.user = user;
+          }
+          resolve(user);
+        });
+      });
+    });
+  }
+
+  /**
+   * Signs in the user to Github using the Firebase API.
+   * @return {!Promise<object>} The logged in user.
+   */
+  signIn() {
+    return firebase.auth().signInWithPopup(this.provider).then(result => {
+      this.accessToken = result.credential.accessToken;
+      this.user = result.user;
+      // A limitation of firebase auth is that it doesn't return an oauth token
+      // after a page refresh. We'll get a firebase token, but not an oauth token
+      // for GH. Since GH's tokens never expire, stash the access token in IDB.
+      return idb.set('accessToken', this.accessToken).then(_ => {
+        return this.accessToken;
+      });
+    });
+  }
+
+  /**
+   * Signs the user out.
+   * @return {!Promise}
+   */
+  signOut() {
+    return firebase.auth().signOut().then(_ => {
+      this.accessToken = null;
+      return idb.delete('accessToken');
+    });
+  }
+}
+
+/**
+ * Wrapper around the Github API for reading/writing gists.
+ * @class
+ */
+class GithubAPI {
+  constructor() {
+    // this.CLIENT_ID = '48e4c3145c4978268ecb';
+    this.auth = new FirebaseAuth();
+  }
+
+  static get LH_JSON_EXT() {
+    return '.lighthouse.report.json';
+  }
+
+  authorize() {
```

Comment:
maybe just absorb this into `FirebaseAuth.signIn`? (this class reaching in to touch `FirebaseAuth.accessToken` is a bit confusing for control flow in general)